craft_application/models/project.py

```diff
@@ -115,6 +124,42 @@ def effective_base(self) -> Any:  # noqa: ANN401 app specific classes can improv
             return self.base
         raise RuntimeError("Could not determine effective base")
 
+    @classmethod
```

Comment:
possibly this should also be `abstract` to force subclasses to implement it 
an "abstract class method" is something that might make sense in python I guess

Reply:
I made it abstract (which required some unit test refactoring).